### DISCUSSION TITLE
saving mp4s

### DISCUSSION THREAD

**Discussion by JulianDay-Cooney** [bonsai-rx/bonsai#919] [Q&A] [answered]
Hi, I want to save long videos (>24 hours) and uncompressed avi format videos are way too large. Is there a way to save them as mp4s?

Thanks!

**Comment by bruno-f-cruz** [comment, marked as answer]
You can use `ffmpeg` to save your videos in a variety of formats. 
[How to install ffmpeg](https://github.com/bonsai-rx/bonsai/discussions/904#discussioncomment-2765294)
To save the videos you can use the attached workflow or you can play around with the recent `VideoWriter(ffmpeg)` node that takes care of all the pipeline for you :)

To save in .mp4, the following settings have worked for me in the past:
`-y -f rawvideo -vcodec rawvideo -s 640x480 -r 30 -pix_fmt bgr24 -i \\.\pipe\video -c:v libx264 -b:v 2M -maxrate 2M -bufsize 1M output.mp4`
[ffmpegmp4.zip](https://github.com/bonsai-rx/bonsai/files/8850918/ffmpegmp4.zip)

**Reply by JulianDay-Cooney** [reply]
Ok I can't get the videowriter to work for me (it keeps saying "runtime error: pipe is broken") when I put in a file path where I want it to be saved, and the output argument as the settings you included. Do you know what the issue may be?

Also, I tried the method in the ffmegmp4 workflow and it is presumably working, but I can't find where the video is being saved to. Do you know where the files are saved and if I can change the save location?

**Reply by bruno-f-cruz** [reply]
It should be saving it in the workflow folder. Otherwise you can define the full path by replacing the "output.mp4"

**Reply by JulianDay-Cooney** [reply]
That worked for the older workflow, but this method does not seem to be compatible with saving multiple video streams (I'm getting the error that "All pipe instances are busy"). Is there a way to make this compatible with more cameras?

Also, any idea why the videowriter isn't working? I'm attaching a screenshot of the input parameters I used. 

![VideoWriterParameters](https://user-images.githubusercontent.com/41309407/172458042-466550af-36ff-43a3-9129-5b2a04c3d379.PNG)

**Reply by bruno-f-cruz** [reply]
If you want to save multiple files simultaneously, you will have to create new pipes and filenames for each one.
As for the ffmpeg node, can you send me your workflow so I can take a look? thanks

**Reply by JulianDay-Cooney** [reply]
Github doesn't seem to like workflow attachments. Any idea how to attach a .bonsai file? But the workflow I'm testing is simply VideoCaptureDevice-->VideoWriter(ffmpeg). The full output argument I use is:
-y -f rawvideo -vcodec rawvideo -s 640x480 -r 30 -pix_fmt bgr24 -i \\.\pipe\video -c:v libx264 -b:v 2M -maxrate 2M -bufsize 1M C:\Users\BehaviorRig\Desktop\LocalData\test.mp4

With regard to creating a new "pipe", I'm not sure exactly what that means. Am I switching out this bit: \\.\pipe\video for a new naming convention?

**Reply by bruno-f-cruz** [reply]
You need to zip the bonsai file, as GitHub only allows a few selected file extensions to be attached unfortunately.

You can change to pipe name by replacing the path after the last "\". Eg. Video2.

**Reply by bruno-f-cruz** [reply]
Sorry I didnt even look at your screenshot, my bad. 
Can you try to use the attached workflow? It should save a `.mp4` from a camera stream.

[FFMPEG_example.zip](https://github.com/bonsai-rx/bonsai/files/8861282/FFMPEG_example.zip)

**Reply by JulianDay-Cooney** [reply]
That worked great! Thanks so much for your help.
